packages/mattermost-redux/src/utils/post_list.ts

```diff
@@ -323,142 +322,100 @@ export function makeGenerateCombinedPost(): (state: GlobalState, combinedId: str
+            // sets the propsUserId to an empty string, this allows the function to combine different "remove" messages with different actors.
+            if (postType === Posts.POST_TYPES.REMOVE_FROM_CHANNEL) {
+                actorId = '';
+            }
+            if (!post.user_id || !post.props) {
+                throw new Error(`Invalid post object: ${JSON.stringify(post)}`);
+            }
             const userId = post.props.addedUserId || post.props.removedUserId;
             const username = post.props.addedUsername || post.props.removedUsername;
-            if (combinedPostType) {
-                if (Array.isArray(combinedPostType[post.user_id])) {
-                    throw new Error('Invalid Post activity data');
-                }
-                const users = combinedPostType[post.user_id] || {ids: [], usernames: []};
-                if (userId) {
-                    if (!users.ids.includes(userId)) {
-                        users.ids.push(userId);
-                    }
-                } else if (username && !users.usernames.includes(username)) {
-                    users.usernames.push(username);
-                }
-                combinedPostType[post.user_id] = users;
-            } else {
-                const users = {
-                    ids: [] as string[],
-                    usernames: [] as string[],
-                };
-
-                if (userId) {
-                    users.ids.push(userId);
-                } else if (username) {
-                    users.usernames.push(username);
-                }
-                userActivityProps[postType] = {
-                    [post.user_id]: users,
-                };
-            }
-        } else {
-            const propsUserId = post.user_id;
 
-            if (combinedPostType) {
-                if (!Array.isArray(combinedPostType)) {
-                    throw new Error('Invalid Post activity data');
-                }
-                if (!combinedPostType.includes(propsUserId)) {
-                    userActivityProps[postType] = [...combinedPostType, propsUserId];
+            // if the post has the same type and userId as the last one, combine them
+            if (prevPost.type === post.type && prevPost.user_id === post.user_id) {
+                // combine userIds and push to userActivities
+                const lastEntry = userActivities[userActivities.length - 1];
+                if (lastEntry) {
+                    lastEntry.userIds?.push(userId);
+                    lastEntry.usernames?.push(username);
                 }
             } else {
-                userActivityProps = {...userActivityProps, [postType]: [propsUserId]};
+                userActivities.push({
+                    actorId,
+                    userIds: [userId],
+                    usernames: [username],
+                    postType,
+                });
             }
+        } else {
+            userActivities.push({
```

Comment:
Since we're not looking at the last post/userActivities here, it's not actually grouping some types of messages
![Screen Shot 2023-03-02 at 6 24 46 PM](https://user-images.githubusercontent.com/3277310/222586419-8ad86d8f-7706-4d2d-9247-759ed10f2f95.png).

Reply:
Perhaps with all this, we could restructure this function as well to reduce the amount of conditional oogic here as well? Maybe we could do something like
```ts
systemPosts.reverse().forEach((post: Post) => {
    const postType = post.type;
    let actorId = post.user_id;
    let userIds;
    let usernames;
    if (isUsersRelatedPost(postType)) {
        actorId = ...;
        userIds = ...;
        usernames = ...;
    } else {
        // Maybe this should also set actorId to blank so that messages like "X and Y have joined the channel" are grouped?
        userIds = ...;
        usernames = ...;
    }

    if (/* postType and actorId are the same as the last entry */) {
        // Add userIds and usernames to the last entry
    } else {
        userActivities.push(/* A new entry */);
    }
});
```